kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/Query.scala

```diff
@@ -0,0 +1,208 @@
+          queryId.operationHandle,
+          defaultFetchOrientation,
+          defaultMaxRows,
+          false)
+        TrinoContext.createQueryResults(
+          queryId.getQueryId,
+          nextUri,
+          queryHtmlUri,
+          status,
+          Option(metaData),
+          Option(resultSet))
+      case _ =>
+        TrinoContext.createQueryResults(
+          queryId.getQueryId,
+          nextUri,
+          queryHtmlUri,
+          status)
+    }
+  }
+
+  def getLastToken: Long = this.lastToken.get()
+
+  def getSlug: Slug = this.slug
+
+  def cancel: Unit = clear
+
+  private def clear = {
+    be.closeOperation(queryId.operationHandle)
+    context.session.get("sessionId").foreach { id =>
+      be.closeSession(SessionHandle.fromUUID(id))
+    }
+  }
+
+  private def setToken(token: Long): Unit = {
+    val lastToken = this.lastToken.get
+    if (token != lastToken && token != lastToken + 1) {
+      throw new WebApplicationException(Response.Status.GONE)
+    }
+    this.lastToken.compareAndSet(lastToken, token)
+  }
+
+  private def getNextUri(token: Long, uriInfo: UriInfo, slugContext: Slug.Context.Context): URI = {
+    val path = slugContext match {
+      case QUEUED_QUERY => QUEUED_QUERY_PATH
+      case EXECUTING_QUERY => EXECUTING_QUERY_PATH
+    }
+
+    uriInfo.getBaseUriBuilder.replacePath(path)
+      .path(queryId.getQueryId)
+      .path(slug.makeSlug(slugContext, token))
+      .path(String.valueOf(token))
+      .replaceQuery("")
+      .build()
+  }
+
+  private def toSlugContext(state: OperationState): Slug.Context.Context = {
+    state match {
+      case INITIALIZED | PENDING => Slug.Context.QUEUED_QUERY
+      case _ => Slug.Context.EXECUTING_QUERY
+    }
+  }
+
+}
+
+object Query {
+
+  def apply(
+      statement: String,
+      context: TrinoContext,
+      translator: KyuubiTrinoOperationTranslator,
+      backendService: BackendService,
+      queryTimeout: Long = 0): Query = {
+
+    val sessionHandle = createSession(context, backendService)
+    val operationHandle = translator.transform(
+      statement,
+      sessionHandle,
+      context.session,
+      true,
+      queryTimeout)
+    val newSessionProperties =
+      context.session + ("sessionId" -> sessionHandle.identifier.toString)
+    val updatedContext = context.copy(session = newSessionProperties)
+    Query(QueryId(operationHandle), updatedContext, backendService)
+  }
+
+  def apply(id: String, context: TrinoContext, backendService: BackendService): Query = {
+    Query(QueryId(id), context, backendService)
+  }
+
+  private def createSession(
+      context: TrinoContext,
+      backendService: BackendService): SessionHandle = {
+    context.session
+      .get("sessionId")
+      .fold(backendService.openSession(
+        TProtocolVersion.HIVE_CLI_SERVICE_PROTOCOL_V11,
+        context.user,
+        "",
+        context.remoteUserAddress.getOrElse(""),
+        context.session))(SessionHandle.fromUUID)
```

Comment:
what does this do ? Trino does not have `session` concept, we can just open a new session for next query with its conext.

Reply:
Yes, this point is also to be further confirmed. Each query corresponds to a new session (per query per session), and then the association between them can be passed through the context. Is this correct?

Reply:
Another question, do we need to actively close the session after the results are fetched?

Reply:
> Each query corresponds to a new session (per query per session), and then the association between them can be passed through the context

yea, because all context is in http header for each query.

> do we need to actively close the session after the results are fetched

What' the behavior of trino-cli ? I remember Trino jdbc driver will send cancel query request when closing connection, then we do not need to eagerly close session.
Besides, Kyuubi session manager has a timeout checker that can close inactive session. So I think we do not need to close session manually.